orca_detector/vggish_model.py

```diff
@@ -0,0 +1,261 @@
+
+class OrcaVGGish(object):    
+    """
+    An adaption of the VGGish architecture for our audio classification model.
+    This model is never trained if weights aren't provided!
+    """
+    
+    def __init__ (self,
+                  load_weights=True, 
+                  weights='audioset', 
+                  input_tensor=None,
+                  input_shape=None,
+                  out_dim=None,
+                  pooling='avg'):
+
+        """
+        :param load_weights: if load weights
+        :param weights: loads weights pre-trained on a preliminary version of YouTube-8M.
+        :param input_tensor: input_layer
+        :param input_shape: input data shape
+        :param out_dim: output dimension
+        :param pooling: pooling type over the non-top network, 'avg' or 'max'
+
+        :return: A Keras model instance.
+        """
+
+        # Validate parameters
+        if weights not in {'audioset', None}:
+            raise ValueError('The `weights` argument should be either '
+                             '`None` (random initialization) or `audioset` '
+                             '(pre-training on audioset).')
+
+        out_dim = orca_params.NUM_CLASSES
+
+        if input_shape is None:
+            input_shape = (params.NUM_FRAMES, params.NUM_BANDS, 1)  # 496, 64, 1
+
+        # VGGish model was trained with a "batch-first" matrix format.
+        if input_tensor is None:
+            aud_input = Input(shape=input_shape, name='input_1')
+        else:
+            if not K.is_keras_tensor(input_tensor):
+                aud_input = Input(tensor=input_tensor, shape=input_shape, name='input_1')
+            else:
+                aud_input = input_tensor
+
+        # TODO: determine if we need a BatchNormalization layer to process input before
+        #   feeding to the pretrained VGGish layers.
+        
+        # Build VGGish model
+        # Block 1
+        x = Conv2D(64, (3, 3), strides=(1, 1), activation='relu', padding='same', name='conv1')(aud_input)
+        x = MaxPooling2D((2, 2), strides=(2, 2), padding='same', name='pool1')(x)
+
+        # Block 2
+        x = Conv2D(128, (3, 3), strides=(1, 1), activation='relu', padding='same', name='conv2')(x)
+        x = MaxPooling2D((2, 2), strides=(2, 2), padding='same', name='pool2')(x)
+
+        # Block 3
+        x = Conv2D(256, (3, 3), strides=(1, 1), activation='relu', padding='same', name='conv3/conv3_1')(x)
+        x = Conv2D(256, (3, 3), strides=(1, 1), activation='relu', padding='same', name='conv3/conv3_2')(x)
+        x = MaxPooling2D((2, 2), strides=(2, 2), padding='same', name='pool3')(x)
+
+        # Block 4
+        x = Conv2D(512, (3, 3), strides=(1, 1), activation='relu', padding='same', name='conv4/conv4_1')(x)
+        x = Conv2D(512, (3, 3), strides=(1, 1), activation='relu', padding='same', name='conv4/conv4_2')(x)
+
+        # FC block (this is the space for our custom adaptations of the VGGish model)
```

Comment:
Does it make sense to break this method into two, the vanila VGGish creation with loaded weights and the 'orca' addition on top of it. I'm thinking of having multiple permutations of the 'orca' part and do a configuration driven dynamic composition, just and idea.

Reply:
Yeah, I can put the generic stuff into its own method to tidy the code up a little.  Trying to generalize this class to support a variety of permutations on top of the VGGish core is one of those cases where I want to be cautious about over-engineering since we don't know yet how many or what type of permutations we'll want to put on top of the core.

I'm more inclined to take your subclasses suggestion, and set up OrcaVGGish to be friendly to being subclassed.  Then rather than having a bunch of if... elif... blocks that add different layers, if you want to experiment with a different model, you'd just subclass OrcaVGGish.  I'll work on this.

Reply:
Sounds good

Reply:
please take a look at the commit I just pushed.  OrcaVGGish is a subclass of VGGish now, and I think it's sufficiently easy to create similar subclasses.